### ISSUE TITLE
Neither CUDA nor NVDEC will enable on Windows

### ISSUE THREAD

**Issue by Doofussy2** [mpv-player/mpv#5978]
Latest Git build, but it's been this way for as long as I've had my GTX 1050 (a few months)
Windows 10

**Reproduction steps**
Using hwdec=cuda or hwdec=nvdec in the mpv.conf, results in software decoding. But cuda-copy and nvdec-copy, work

**Expected behavior**
Hardware decoding

[Portable mpv log.txt](https://github.com/mpv-player/mpv/files/2174215/Portable.mpv.log.txt)

I've also reported this Nvidia, in case it's a driver issue.

**Comment by morphological** [comment]
Encoding works fine when ffmpeg is compiled with --enable-nvenc and --enable-cuvid 
Gotta install the headers first though: https://github.com/lutris/ffmpeg-nvenc/issues/22
Guide  over here: https://github.com/mpv-player/mpv-build
```mpv --ovc=help | grep nvenc
  --ovc=h264_nvenc   NVIDIA NVENC H.264 encoder
  --ovc=nvenc        NVIDIA NVENC H.264 encoder
  --ovc=nvenc_h264   NVIDIA NVENC H.264 encoder
  --ovc=nvenc_hevc   NVIDIA NVENC hevc encoder
  --ovc=hevc_nvenc   NVIDIA NVENC hevc encoder
```
Decoding with nvenc also works btw.

**Comment by Doofussy2** [comment]
Thanks for responding, but this is confusing. First, I'm not encoding anything. Second, the mpv manual says to use hwdec=cuda with vo=gpu. This is the method used for all hwdec, in the manual. It works with other decoders but not cuda or nvdec. Has something changed that hasn't been updated in the manual?

**Comment by aufkrawall** [comment]
cuda & nvdec only work with --gpu-api=opengl

**Comment by Doofussy2** [comment]
I believe I tried that, with no success. I will retest, when I get a chance.

Thank you.

**Comment by Doofussy2** [comment]
Just tested with gpu-api=opengl with cuda and nvdec. No change. They will not enable. Other guys I've talked to online, also can't get them to enable.

**Comment by aufkrawall** [comment]
You're right, doesn't work here either with @shinchiro 's build while it works with self compiled build on Linux.

**Comment by saildata** [comment]
Works here. https://i.imgur.com/bOWgOJb.jpg

OS: Arch Linux 4.17.5, 1080-Ti 11G

- Installed ffnvcodec-headers as others have mentioned. 
- Added `--enable-nvenc` to ffmpeg build options.

`mpv --version`
    
```mpv 0.28.0-633-g351c083487 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       56.18.102
   libavcodec      58.21.105
   libavformat     58.17.101
   libswscale      5.2.100
   libavfilter     7.25.100
   libswresample   3.2.100
ffmpeg version: N-91473-gc5329d64b1```

MPV stdout:
 ```(+) Video --vid=1 (*) (vp9 3840x2160 60.000fps)
 (+) Audio --aid=1 --alang=eng (*) (opus 2ch 48000Hz)
Using hardware decoding (nvdec).
AO: [pulse] 48000Hz stereo 2ch float
VO: [gpu] 3840x2160 cuda[nv12]
```

I also have ```vo=gpu
hwdec=auto
gpu-api=auto```
it was not necessary to have `gpu-api=opengl` for mine. Just for fun, I ran the same video in 'benchmark' mode:
    
```
[bench]
hwdec=cuda
scale=ewa_lanczossharp
cscale=ewa_lanczossharp
vd-lavc-threads=32
audio=no
untimed=yes
video-sync=display-desync
vulkan-swap-mode=immediate
opengl-swapinterval=0
opengl-pbo
osd-msg1="FPS: ${estimated-display-fps}"
```
    
    
and got around 250fps on 3840*2160 UHD. All seems to be fine on this end. I think it just depends on your build and having the right libraries installed. 
https://i.imgur.com/YIaezFe.jpg

**Comment by shinchiro** [comment]
Uh, I'll explicitly add `--enable-nvdec` when configuring ffmpeg if that will change anything. It's supposed to be on (autodetect) by default before.

**Comment by saildata** [comment]
I think I also used `--enable-cuvid` or `--enable-cuda` I can't remember which. Sorry I built in `/tmp` and deleted my makefiles.

**Comment by Doofussy2** [comment]
@shinchiro , unfortunately, cuda nor nvdec are still not enabling in your latest build :(

**Comment by shinchiro** [comment]
```
[vo/gpu] Loading hwdec driver 'cuda-nvdec'
Cannot load cuInit
[vo/gpu/cuda-nvdec] Failed to load CUDA symbols
[vo/gpu] Loading failed.
[vd] Could not create device.
```
Just realized, it's always enable before. I believe you need to download/install cuda toolkit (~1.5GB) to use it.

**Comment by Doofussy2** [comment]
@shinchiro . I installed it, no change. It still doesn't enable.

**Comment by aufkrawall** [comment]
I don't think that something cuda/nvdec related is missing, as -copy modes work. It rather seems to me that something interop related doesn't work with Windows builds.

**Comment by Doofussy2** [comment]
This is what I see in the log

> 
[   0.233][v][vd] Codec list:
[   0.233][v][vd]     hevc - HEVC (High Efficiency Video Coding)
[   0.233][v][vd]     hevc_cuvid (hevc) - Nvidia CUVID HEVC decoder
[   0.233][v][vd] Opening decoder hevc
[   0.234][v][vd] Looking at hwdec hevc-nvdec...
[   0.234][v][vo/gpu] Loading hwdec driver 'd3d11-egl'
[   0.235][v][vo/gpu] Loading failed.
[   0.235][v][vo/gpu] Loading hwdec driver 'd3d11-egl-rgb'
[   0.235][v][vo/gpu] Loading failed.
[   0.235][v][vo/gpu] Loading hwdec driver 'dxva2-egl'
[   0.235][v][vo/gpu] Loading failed.
[   0.235][v][vo/gpu] Loading hwdec driver 'd3d11va'
[   0.240][v][vo/gpu] Loading hwdec driver 'dxva2-dxgi'
[   0.338][v][vo/gpu] Loading hwdec driver 'dxva2-dxinterop'
[   0.338][v][vo/gpu] Loading failed.
[   0.338][v][vo/gpu] Loading hwdec driver 'cuda-nvdec'
[   0.338][v][vo/gpu] Loading failed.
[   0.338][v][vd] Could not create device.
[   0.338][v][vd] No hardware decoding available for this codec.
[   0.338][v][vd] Using software decoding.

**Comment by Doofussy2** [comment]
It isn't enabling in 0.29, either

**Comment by g3r** [comment]
This is not a Windows only problem. I have the same problem in archlinux, reverted to 0.28 where it works fine.

**Comment by Doofussy2** [comment]
Not to be a pain, but is there any development on on this? Is this an OS issue?

**Comment by tomty89** [comment]
@Doofussy2 Maybe you can set the message level to trace to see at which function it failed exactly:
https://github.com/mpv-player/mpv/blob/07915b12273a36bc7f104a5f3fc949a407d243dc/video/out/opengl/hwdec_cuda.c#L61

**Comment by Doofussy2** [comment]
Thanks @tomty89 I'm going to add --msg-level=all=trace to the mpv.conf, then post the log

**Comment by Doofussy2** [comment]
Thanks @tomty89. Here's the log with that, using @shinchiro's build

[Portable mpv log.txt](https://github.com/mpv-player/mpv/files/2234324/Portable.mpv.log.txt)

**Comment by tomty89** [comment]
Somehow the message level was not increased in this part:
```
[   0.169][v][vd] Looking at hwdec hevc_cuvid-cuda...
[   0.169][v][vo/gpu] Loading hwdec driver 'd3d11-egl'
[   0.169][v][vo/gpu] Loading failed.
[   0.169][v][vo/gpu] Loading hwdec driver 'd3d11-egl-rgb'
[   0.169][v][vo/gpu] Loading failed.
[   0.169][v][vo/gpu] Loading hwdec driver 'dxva2-egl'
[   0.169][v][vo/gpu] Loading failed.
[   0.169][v][vo/gpu] Loading hwdec driver 'd3d11va'
[   0.170][v][vo/gpu] Loading hwdec driver 'dxva2-dxgi'
[   0.216][v][vo/gpu] Loading hwdec driver 'dxva2-dxinterop'
[   0.216][v][vo/gpu] Loading failed.
[   0.216][v][vo/gpu] Loading hwdec driver 'cuda-nvdec'
[   0.216][v][vo/gpu] Loading failed.
[   0.216][v][vd] Could not create device.
```
Maybe we need `--msg-level=vd=trace` instead? Not really sure how `--msg-level` works to be honest.

**Comment by Doofussy2** [comment]
That appears to be giving the same output, but here it is

[Portable mpv log.txt](https://github.com/mpv-player/mpv/files/2234391/Portable.mpv.log.txt)

**Comment by tomty89** [comment]
Hmm what about `--msg-level=vo/cuda-nvdec=strace` or `--msg-level=vo=strace`?

**Comment by Doofussy2** [comment]
I think what is probably required, is this ---msg-level=vo/gpu=trace. Which makes this log. I'm not sure it provides any more info, though.

[Portable mpv log.txt](https://github.com/mpv-player/mpv/files/2234495/Portable.mpv.log.txt)

**Comment by sc44** [comment]
Hello,
i am using Arch too. I have recompiled mpv 0.29  on my System with the community PKGBUILD.
Now it runs. I'm not sure if that's important, i use my own ffmpeg version.
https://pastebin.com/StVSjHAN

**Comment by Doofussy2** [comment]
Is there anything else I can provide, to help?

**Comment by DistantThunder** [comment]
Just popping to say that I still observe this behavior on Arch Linux with:

> mpv --version
> 
> mpv 0.29.0 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
>  built on Sun Jul 22 21:06:23 CEST 2018
> ffmpeg library versions:
>    libavutil       56.14.100
>    libavcodec      58.18.100
>    libavformat     58.12.100
>    libswscale      5.1.100
>    libavfilter     7.16.100
>    libswresample   3.1.100
> ffmpeg version: n4.0.2

[...]
```
[cplayer]  (+) Video --vid=1 (*) (h264 1920x1080 59.940fps)
[cplayer]  (+) Audio --aid=1 --alang=eng (*) 'DASH audio' (aac 2ch 44100Hz) (external)
[vo/gpu] Probing for best GPU context.
[vo/gpu/opengl] Initializing GPU context 'x11probe'
[vo/gpu/x11] X11 opening display: :0
[vo/gpu/x11] X11 running at 1920x1080 (":0" => local display)
[vo/gpu/x11] Detected wm supports NetWM.
[vo/gpu/x11] Detected wm supports FULLSCREEN state.
[vo/gpu/x11] Detected wm supports ABOVE state.
[vo/gpu/x11] Detected wm supports BELOW state.
[vo/gpu/x11] Display 0 (HDMI-0): [0, 0, 1920, 1080] @ 60.000000 FPS
[vo/gpu/x11] Current display FPS: 60.000000
[vo/gpu/opengl] GLX chose FB config with ID 0xad
[vo/gpu/opengl] GLX chose visual with ID 0x27
[vo/gpu/opengl] Creating OpenGL 4.4 context...
[vo/gpu] GL_VERSION='4.4.0 NVIDIA 396.45'
[vo/gpu] Detected desktop OpenGL 4.4.
[vo/gpu] GL_VENDOR='NVIDIA Corporation'
[vo/gpu] GL_RENDERER='GeForce GTX 1060 6GB/PCIe/SSE2'
[vo/gpu] GL_SHADING_LANGUAGE_VERSION='4.40 NVIDIA via Cg compiler'
[vo/gpu] Loaded extension GLX_SGI_swap_control.
[vo/gpu] Loaded extension GLX_SGI_video_sync.
[vo/gpu] Loaded extension GL_NV_vdpau_interop.
[vo/gpu] Testing FBO format rgba16f
[vo/gpu] Using FBO format rgba16f.
[vo/gpu] Assuming 60.000000 FPS for display sync.
[vd] Container reported FPS: 59.940060
[vd] Codec list:
[vd]     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
[vd]     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper
[vd]     h264_cuvid (h264) - Nvidia CUVID H264 decoder
[vd] Opening decoder h264
[vd] Looking at hwdec h264-nvdec...
[vo/gpu] Loading hwdec driver 'vaapi-egl'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'vdpau-glx'
[vo/gpu] Loading hwdec driver 'drmprime-drm'
[vo/gpu/drmprime-drm] Failed to retrieve DRM fd from native display.
[vo/gpu] Loading failed.
[vd] Could not create device.
[vd] No hardware decoding available for this codec.
[vd] Using software decoding.
[vd] Detected 16 logical cores.
[vd] Requesting 16 threads for decoding.
[vd] Selected codec: h264 (H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10)
[vf] User filter list:
[vf]   (empty)

```
[...]


Which is strange because ffmpeg itself has no problem using cuvid API: 

```

ffplay -v debug -codec:v h264_cuvid Vidéos/Halo\ 5\ Opening\ Cinematic.mp4 ^| grep cuvid                                                     
[h264_cuvid @ 0x7f8144015880] Format nv12 chosen by get_format().
[h264_cuvid @ 0x7f8144015880] Loaded lib: libnvcuvid.so.1
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidGetDecoderCaps
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCreateDecoder
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidDestroyDecoder
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidDecodePicture
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidMapVideoFrame64
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidUnmapVideoFrame64
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCtxLockCreate
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCtxLockDestroy
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCtxLock
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCtxUnlock
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCreateVideoSource
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCreateVideoSourceW
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidDestroyVideoSource
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidSetVideoSourceState
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidGetVideoSourceState
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidGetSourceVideoFormat
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidGetSourceAudioFormat
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidCreateVideoParser
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidParseVideoData
[h264_cuvid @ 0x7f8144015880] Loaded sym: cuvidDestroyVideoParser
[h264_cuvid @ 0x7f8144015880] CUVID capabilities for h264_cuvid:
[h264_cuvid @ 0x7f8144015880] 8 bit: supported: 1, min_width: 48, max_width: 4096, min_height: 16, max_height: 4096
[h264_cuvid @ 0x7f8144015880] 10 bit: supported: 0, min_width: 0, max_width: 0, min_height: 0, max_height: 0
[h264_cuvid @ 0x7f8144015880] 12 bit: supported: 0, min_width: 0, max_width: 0, min_height: 0, max_height: 0
[h264_cuvid @ 0x7f8144015880] Format nv12 chosen by get_format().
[h264_cuvid @ 0x7f8144015880] Formats: Original: nv12 | HW: nv12 | SW: nv12
```


I also wanted to say that as suggested in this thread, I recompiled v0.29 from git locally but the same behavior still happens. 

Tried using latest Git (0.29.0-20-gd8131568c8) but this is still an issue.

Reverting to distribution's 0.28.2 build, nvdec is functional again.

**Comment by DistantThunder** [comment]
After some amateur digging, I believe commit 07915b12273a36bc7f104a5f3fc949a407d243dc is where things went funky.

Now that Cuda loader custom logic has been dropped out of MPV, it relies on the `ffnvcodec-headers` separate packages to support CUDA related API codecs in FFMPEG. 

These headers have to be installed at build time or else MPV runtime won't be able to use FFMPEG's implementation of nvdec/cuvid. 

I believe Arch Linux's current MPV build do not include this dependency as it is typically an extra feature that won't break decoding in any major way for most installs but far from me to be guessing what went on. It's simply possible most maintainers didn't realize the change and didn't notice the new separate headers. 

Anyhow, this also explains why some like myself kept seeing the problem even with local builds while those fixed other people situation: it all depends on whether you have the `ffnvcodec-headers` package when building MPV. 

### TL;DR

1. Install `ffnvcodec-headers` package on your distro or obtain the headers in whichever preferred way.
2. Recompile MPV.
3. ???
4. Profit from HW decoding again on NVIDIA platforms.

- Extra-step: Bump the maintainers of your distro about this issue. Even if MPV itself is compiled with ffnvcodec support, the end-users won't have to have the headers installed. The runtime will be perfectly. This adds extra dependency at build time (maintainers's side) but requires no extra dependency for end-users. I may not know all the constraints for maintainers, but it appears to me this shouldn't be a problem.

**Comment by philipl** [comment]
Yes, this is intended behaviour. You need the headers available at build time. Whoever has rights to close this issue can do so.

**Comment by Doofussy2** [comment]
So is this now removed from the release builds? And only copy modes can be used?

**Comment by philipl** [comment]
No. You need the headers available at build time as noted by @DistantThunder.

**Comment by Doofussy2** [comment]
Right, but @lachs0r is the one doing the build. So my question remains the same. Is it being removed?

**Comment by shinchiro** [comment]
@philipl Is it possible there's cuda interop problem in windows? My ffmpeg can decode with cuda by running: `ffmpeg -hwaccel cuda -threads 1 -i INPUT  -f null - -benchmark` . While configure mpv, I also set `--enable-cuda-hwaccel` to confirm it was detected. But I still get the error, "Cannot load cuInit"

**Comment by philipl** [comment]
@shinchiro I'm not in a position to test on windows, but the loading code, assuming you built with the right nvidia headers (https://github.com/FFmpeg/nv-codec-headers) is the same as used in ffmpeg, so if one can load the symbols, the other should be able to.

Really the only thing that can go wrong there is not finding the dll, but again, if ffmpeg can find it, mpv should be able to as well.

**Comment by aufkrawall** [comment]
It's btw. fixed with latest repo build on Arch.

**Comment by prowal** [comment]
I did some research and can't find any windows build with nvdec or cuda actually working (recent or past @lachs0r or @shinchiro builds)

In recent  builds, the issue doesn't seem to be the same:

 @lachs0r:
[vd] Codec list:
[vd]     hevc - HEVC (High Efficiency Video Coding)
[vd]     hevc_cuvid (hevc) - Nvidia CUVID HEVC decoder
[vd] Opening decoder hevc
[vd] Looking at hwdec hevc_cuvid-cuda...
[vo/gpu] Loading hwdec driver 'd3d11va'
[vo/gpu] Loading hwdec driver 'dxva2-dxgi'
[vo/gpu] Loading hwdec driver 'dxva2-dxinterop'
[vo/gpu] Loading failed.
[vd] Could not create device.
[vd] No hardware decoding available for this codec.
[vd] Using software decoding.

 @shinchiro:

[vd] Codec list:
[vd]     hevc - HEVC (High Efficiency Video Coding)
[vd]     hevc_cuvid (hevc) - Nvidia CUVID HEVC decoder
[vd] Opening decoder hevc
[vd] Looking at hwdec hevc_cuvid-cuda...
[vo/gpu] Loading hwdec driver 'd3d11-egl'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'd3d11-egl-rgb'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'dxva2-egl'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'd3d11va'
[vo/gpu] Loading hwdec driver 'dxva2-dxgi'
[vo/gpu] Loading hwdec driver 'dxva2-dxinterop'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'cuda-nvdec'
[vo/gpu] Loading failed.
[vd] Could not create device.
[vd] No hardware decoding available for this codec.
[vd] Using software decoding.

**Comment by philipl** [comment]
Those lines say that @lachs0r build doesn't have the cuda hwdec built into it but the @shinchiro ones does. So you need to focus on the @shinchiro one.

You need to turn the logging all the way up to TRACE and see what else it says when it tries to load it. I would expect it's saying it can't find the right dlls. Please grab the additional logs and add them here.

**Comment by prowal** [comment]
there isn't more information with --msg-level=all=trace:
[vd] Codec list:
[vd]     hevc - HEVC (High Efficiency Video Coding)
[vd]     hevc_cuvid (hevc) - Nvidia CUVID HEVC decoder
[vd] Opening decoder hevc
[vd] Looking at hwdec hevc_cuvid-cuda...
[vo/gpu] Loading hwdec driver 'd3d11-egl'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'd3d11-egl-rgb'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'dxva2-egl'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'd3d11va'
[vo/gpu] Loading hwdec driver 'dxva2-dxgi'
[vo/gpu] Loading hwdec driver 'dxva2-dxinterop'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'cuda-nvdec'
[vo/gpu] Loading failed.
[vd] Could not create device.
[vd] No hardware decoding available for this codec.
[vd] Using software decoding.


BUT: i realized that hwdecode is **actually working** with --hwdec=auto:
[vd] Codec list:
[vd]     hevc - HEVC (High Efficiency Video Coding)
[vd]     hevc_cuvid (hevc) - Nvidia CUVID HEVC decoder
[vd] Opening decoder hevc
[vd] Looking at hwdec hevc-d3d11va...
[vo/gpu] Loading hwdec driver 'd3d11-egl'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'd3d11-egl-rgb'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'dxva2-egl'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'd3d11va'
[vo/gpu] Loading hwdec driver 'dxva2-dxgi'
[vo/gpu] Loading hwdec driver 'dxva2-dxinterop'
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'cuda-nvdec'
[vo/gpu] Loading failed.
[vd] Trying hardware decoding via hevc-d3d11va.
[ffmpeg/video] hevc: Invalid default display window
[vd] Selected codec: hevc (HEVC (High Efficiency Video Coding))

**Comment by Doofussy2** [comment]
I've tried trace logs in previous posts, but it doesn't reveal any more information. Is there a specific expression to expand the GPU logging?

**Comment by Doofussy2** [comment]
Is it not generating more because it isn't being enabled, or is it actually the logging that needs augmenting?

**Comment by Doofussy2** [comment]
Interesting. Maybe Linux has more tools to be able to analyze more thoroughly?

**Comment by shinchiro** [comment]
@philipl on my windows system, it failed at https://github.com/mpv-player/mpv/blob/104b51077451a3f9e259ca9aeff0c70ac1aa1d27/video/out/opengl/hwdec_cuda.c#L100
so the `CHECK_CU` function never called in the first place. Of course, the dlls is there as ffmpeg can find it

**Comment by shinchiro** [comment]
Here's the extracted symbols found in nvcuda.dll: https://0x0.st/sELa.txt

I did compared the symbols with the one in `nv-codec-headers` and some of the symbols do not exist in the dll. Probably that's fault, I guess?

Here's the symbols lines which absent in the dll
https://github.com/FFmpeg/nv-codec-headers/blob/d098cb44f6103a03476d16ac2bbe14ecbc9e1175/include/ffnvcodec/dynlink_loader.h#L175-L178

**Comment by shinchiro** [comment]
yeah, `nvdec-copy` works
```
[vd] Codec list:
[vd]     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
[vd]     h264_cuvid (h264) - Nvidia CUVID H264 decoder
[vd] Opening decoder h264
[vd] Looking at hwdec h264-nvdec-copy...
[ffmpeg] AVHWDeviceContext: Loaded lib: nvcuda.dll
[ffmpeg] AVHWDeviceContext: Loaded sym: cuInit
[ffmpeg] AVHWDeviceContext: Loaded sym: cuDeviceGetCount
[ffmpeg] AVHWDeviceContext: Loaded sym: cuDeviceGet
[ffmpeg] AVHWDeviceContext: Loaded sym: cuDeviceGetName
[ffmpeg] AVHWDeviceContext: Loaded sym: cuDeviceGetUuid
[ffmpeg] AVHWDeviceContext: Loaded sym: cuDeviceComputeCapability
[ffmpeg] AVHWDeviceContext: Loaded sym: cuCtxCreate_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuCtxSetLimit
[ffmpeg] AVHWDeviceContext: Loaded sym: cuCtxPushCurrent_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuCtxPopCurrent_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuCtxDestroy_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuMemAlloc_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuMemFree_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuMemcpy2D_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuMemcpy2DAsync_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGetErrorName
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGetErrorString
[ffmpeg] AVHWDeviceContext: Loaded sym: cuStreamCreate
[ffmpeg] AVHWDeviceContext: Loaded sym: cuStreamQuery
[ffmpeg] AVHWDeviceContext: Loaded sym: cuStreamSynchronize
[ffmpeg] AVHWDeviceContext: Loaded sym: cuStreamDestroy_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuStreamAddCallback
[ffmpeg] AVHWDeviceContext: Loaded sym: cuEventCreate
[ffmpeg] AVHWDeviceContext: Loaded sym: cuEventDestroy_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuEventSynchronize
[ffmpeg] AVHWDeviceContext: Loaded sym: cuEventQuery
[ffmpeg] AVHWDeviceContext: Loaded sym: cuEventRecord
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGLGetDevices_v2
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGraphicsGLRegisterImage
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGraphicsUnregisterResource
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGraphicsMapResources
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGraphicsUnmapResources
[ffmpeg] AVHWDeviceContext: Loaded sym: cuGraphicsSubResourceGetMappedArray
[ffmpeg] AVHWDeviceContext: Loaded sym: cuImportExternalMemory
[ffmpeg] AVHWDeviceContext: Loaded sym: cuDestroyExternalMemory
[ffmpeg] AVHWDeviceContext: Loaded sym: cuExternalMemoryGetMappedBuffer
[ffmpeg] AVHWDeviceContext: Loaded sym: cuExternalMemoryGetMappedMipmappedArray
[ffmpeg] AVHWDeviceContext: Loaded sym: cuMipmappedArrayGetLevel
[vd] Trying hardware decoding via h264-nvdec-copy.
```

The error before came from hwdec `nvdec` or `cuda` (without copy).

**Comment by philipl** [comment]
Ok, I went and dug out my windows installation and updated drivers, etc, etc.

The first observation for everyone who isn't @shinchiro is you must set your --gpu-api=opengl and you must make sure it uses the 'win' context rather than angle. Interop only works with the real system opengl and angle wraps directx.

Based on the errors I saw, I think this is why people don't see any error messages.

When I switched my backend and context correct, I saw what @shinchiro saw - nvdec-copy works with symbol loading and nvdec does not. In my case I also saw a specific error (I think it prints to stdout which is why @shinchiro may have missed it) that said it can't load cuInit.

That basically must mean it can't find nvcuda.dll or it's finding the wrong one. The dll finding logic is in the headers and should be the same in both ffmpeg and mpv. So that's weird.

@BtbN, any thoughts?

**Comment by philipl** [comment]
Continuing. the headers do symbol loading in a simplified way vs ffmpeg, and this is probably where the difference comes from. I guess it's too simple to actually work on windows.

**Comment by philipl** [comment]
@BtbN found and fixed the problem.

https://github.com/FFmpeg/nv-codec-headers/commit/5054a06eb730fa5f024b945d62b13a30ab2d42d8

@shinchiro if you rebuild with the new nv-codec-headers, it should start working.

**Comment by shinchiro** [comment]
Yeah, that fixed it 👍

**Comment by BtbN** [comment]
I'm genuinely baffled how this did not cause a single warning.
Would have expected a huge bunch of compile errors, but no, nothing.

**Comment by Doofussy2** [comment]
Thanks for figuring it out, guys. I look forward to testing it

**Comment by Doofussy2** [comment]
I'm a little confused. First, nvdec is working as expected. But I'm not sure cuda, is.
[Portable mpv log.txt](https://github.com/mpv-player/mpv/files/2498425/Portable.mpv.log.txt)

This is what I see in the log

[   0.413][v][vd] Opening decoder hevc
[   0.413][v][vd] Looking at hwdec hevc_cuvid-cuda...
[   0.413][v][vo/gpu] Loading hwdec driver 'd3d11-egl'
[   0.413][v][vo/gpu] Loading failed.
[   0.413][v][vo/gpu] Loading hwdec driver 'd3d11-egl-rgb'
[   0.413][v][vo/gpu] Loading failed.
[   0.413][v][vo/gpu] Loading hwdec driver 'dxva2-egl'
[   0.413][v][vo/gpu] Loading failed.
[   0.413][v][vo/gpu] Loading hwdec driver 'd3d11va'
[   0.413][v][vo/gpu] Loading failed.
[   0.413][v][vo/gpu] Loading hwdec driver 'dxva2-dxgi'
[   0.413][v][vo/gpu] Loading failed.
[   0.413][v][vo/gpu] Loading hwdec driver 'dxva2-dxinterop'
[   0.413][v][vo/gpu] Loading failed.
[   0.413][v][vo/gpu] Loading hwdec driver 'cuda-nvdec'
[   0.549][v][vd] Trying hardware decoding via hevc_cuvid-cuda.
[   0.549][v][vd] Using underlying hw-decoder 'hevc_cuvid'
[   0.549][v][vd] Pixel formats supported by decoder: cuda nv12
[   0.549][v][vd] Codec profile: unknown (0xffffff9d)
[   0.549][v][vd] Requesting pixfmt 'cuda' from decoder.
[   0.551][d][ffmpeg/video] hevc_cuvid: CUVID capabilities for hevc_cuvid:
[   0.551][d][ffmpeg/video] hevc_cuvid: 8 bit: supported: 1, min_width: 144, max_width: 8192, min_height: 144, max_height: 8192
[   0.551][d][ffmpeg/video] hevc_cuvid: 10 bit: supported: 1, min_width: 144, max_width: 8192, min_height: 144, max_height: 8192
[   0.551][d][ffmpeg/video] hevc_cuvid: 12 bit: supported: 1, min_width: 144, max_width: 8192, min_height: 144, max_height: 8192
[   0.553][v][vd] Selected codec: hevc (HEVC (High Efficiency Video Coding))

And the stats says it isn't.

![snapshot_66](https://user-images.githubusercontent.com/15950919/47259457-a086c000-d45e-11e8-949b-032632b907a5.jpg)

**Comment by Doofussy2** [comment]
That's not true. I get hwdec with nvdec, nvdec-copy, cuda-copy, auto-copy, d3d11va and d3d11va-copy. cuda is the only on that fails.

**Comment by Doofussy2** [comment]
I think I see why. I was starting playback with --fs. If I start playback without that, cuda is enabled.

**Comment by philipl** [comment]
Looking at your full log, yes, what I see is it trying to use cuda and then that failing. Specifically:

```
[   0.815][d][vo/gpu] Resizing texture: 3840x1604
[   0.828][e][ffmpeg/video] hevc_cuvid: av_hwframe_get_buffer failed
[   0.828][w][vd] Error while decoding frame!
[   0.839][d][ao/wasapi] Thread Reset
[   0.851][e][ffmpeg/video] hevc_cuvid: av_hwframe_get_buffer failed
[   0.851][w][vd] Error while decoding frame!
[   0.861][e][ffmpeg/video] hevc_cuvid: av_hwframe_get_buffer failed
[   0.861][w][vd] Error while decoding frame!
[   0.861][e][ffmpeg/video] hevc_cuvid: av_hwframe_get_buffer failed
[   0.861][w][vd] Error while decoding frame!
[   0.946][w][vd] Falling back to software decoding.
```

I can somewhat imagine this is related to --fs being used, but I'm surprised to hear it.

Note that the difference between `cuda` and `nvdec` is that `cuda` uses the nvidia parser and codec implementation while `nvdec` plugs the decoder into the ffmpeg codec using the hwaccel mechanism. Differences in parser and codec features can produce different results (usually the nvidia stuff is more limited than the ffmpeg equivalents), and ultimately the only reason to use `cuda` over `nvdec` is that you want to do deinterlacing, which is only possible with `cuda`.

**Comment by Doofussy2** [comment]
Yeah, I used to enable cuda/cuda-copy for the deinterlacing feature, but then discovered that cuda/cuda-copy can't decode half of my HDR stuff. The picture breaks up, very badly, and freezes. All other decoders are just fine. So I won't be using cuda, I just thought I'd bring it to your attention, in case there was problem. It definitely doesn't work with --fs. Thanks for all the good work, guys.

**Comment by philipl** [comment]
There definitely seem to be nvidia parser/codec issues with HDR stuff. I've seen it too, but fortunately there's no interlaced HDR :-) I'm glad it's working reasonably now.

**Comment by Doofussy2** [comment]
Ha! Yeah, I've had some major issues with deinterlacing. I've had to make auto-profiles, to work around it.

Thanks again

**Comment by Doofussy2** [comment]
I hope @lachs0r will add this to the next release :)
